File: _5_longest_palindromic_substring.java
/*
5. Longest Palindromic Substring

    Total Accepted: 131355
    Total Submissions: 550707
    Difficulty: Medium

Given a string S, find the longest palindromic substring in S.
You may assume that the maximum length of S is 1000,
and there exists one unique longest palindromic substring.
*/
public class Solution {
    public String longestPalindrome(String s) {
        char[] sc = s.toCharArray();
        int length = 0;
        StringBuilder longest = new StringBuilder();
        for (int i = 0; i < sc.length; i++) {
            int r = i, l = i, r2 = i, l2 = i;
            if ((r+(length/2)) > sc.length) break;
            // check even
            if (i+1 < sc.length && sc[i] == sc[i+1]) {
                r++;
                while (l-1 >= 0 && r+1 < sc.length && sc[r+1] == sc[l-1]) {
                    r++;
                    l--;
                }
            }
            while (l2-1 >= 0 && r2+1 < sc.length && sc[r2+1] == sc[l2-1]) {
                r2++;
                l2--;
            }
            if (r2 > r || l2 < l) {
                r = r2;
                l = l2;
            }
            if ((r-l+1) > length) {
                length = r-l+1;
                longest = new StringBuilder();
                for (int j = l; j <= r; j++)
                    longest.append(sc[j]);
            }
        }
        return longest.toString();
    }
}

File: 5/_5_longest_palindromic_substring.py
"""
5. Longest Palindromic Substring

    Total Accepted: 131355
    Total Submissions: 550707
    Difficulty: Medium

Given a string S, find the longest palindromic substring in S.
You may assume that the maximum length of S is 1000,
and there exists one unique longest palindromic substring.
"""
class Solution(object):
    def longestPalindrome(self, s):
        """
        :type s: str
        :rtype: str
        """
        length, longest = 0, ""
        for i in xrange(len(s)):
            #print i,
            r = l = i
            if r+(length/2) > len(s):
                break
            # check even
            if s[i] == s[i-1]:
                l -= 1
                while l >= 0 and r < len(s) and s[l] == s[r]:
                    r, l = r+1, l-1
                #print s[l:r+1]
            # check odd
            r2 = l2 = i
            while l2 >= 0 and r2 < len(s) and s[l2] == s[r2]:
                r2, l2 = r2+1, l2-1
            #print s[l2:r2+1]
            if r2 > r:
                r, l = r2, l2
            if r-l+1 > length:
                length, longest = r-l+1, s[l+1:r]

        return longest
